pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+	pub struct Order<OrderId, AccountId, AssetId, Balance> {
+		order_id: OrderId,
+		placing_account: AccountId,
+		asset_in_id: AssetId,
+		asset_out_id: AssetId,
+		sell_amount: Balance,
+		price: Balance,
+	}
+
+	#[derive(Clone, Copy, Debug, Encode, Decode, Eq, PartialEq, MaxEncodedLen, TypeInfo)]
+	pub struct Claim<T: Config> {
+		claiming_account: T::AccountId,
+		order_claiming: T::Hash,
+	}
+
+	#[pallet::storage]
+	pub type Orders<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		T::Hash,
+		Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::Balance>,
+		ResultQuery<Error<T>::OrderNotFound>,
+	>;
+
+	/// Map of orders for a particular user
+	/// Used to query orders for a particular user using the
+	/// account id of the user as prefix
+	#[pallet::storage]
+	pub type UserOrders<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::AccountId,
+		Twox64Concat,
+		T::Hash,
+		Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::Balance>,
+		ResultQuery<Error<T>::OrderNotFound>,
+	>;
+
+	/// Stores Nonce for orders placed
+	/// Given that Nonce is to ensure that all orders have a unique ID, we can
+	/// use just one Nonce, which means that we only have one val in storage,
+	/// and we don't have to insert new map values upon a new account/currency
+	/// order creation.
+	#[pallet::storage]
+	pub type NonceStore<T: Config> = StorageValue<_, T::Nonce, ValueQuery>;
+	#[pallet::storage]
+	pub type AssetPairOrders<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::AssetCurrencyId,
+		Twox64Concat,
+		T::AssetCurrencyId,
+		BoundedVec<T::Hash, ConstU32<1_000_000>>,
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AssetPairOrdersOverflow,
+		ConflictingAssetIds,
+		InsufficientAssetFunds,
+		InsufficientReserveFunds,
+		InvalidAssetId,
+		OrderNotFound,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		<T as frame_system::Config>::Hash: PartialEq<<T as frame_system::Config>::Hash>,
+	{
+		#[pallet::call_index(0)]
+		// dummy weight for now
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 2).ref_time())]
+		pub fn create_order(
+			origin: OriginFor<T>,
+			asset_in: T::AssetCurrencyId,
+			asset_out: T::AssetCurrencyId,
+			amount: T::Balance,
+			price: T::Balance,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			ensure!(asset_in != asset_out, Error::<T>::ConflictingAssetIds);
+			ensure!(
+				T::AssetRegistry::metadata(&asset_in).is_some(),
+				Error::<T>::InvalidAssetId
+			);
+			ensure!(
+				T::AssetRegistry::metadata(&asset_out).is_some(),
+				Error::<T>::InvalidAssetId
+			);
+			ensure!(
+				T::TradeableAsset::can_reserve(asset_in, &account_id, amount),
+				Error::<T>::InsufficientAssetFunds,
+			);
+
+			ensure!(
+				T::ReserveCurrency::can_reserve(
+					&account_id,
+					T::Fees::fee_value(T::OrderFeeKey::get())
```

Comment:
Could be a conflict if we allowed `trading_curreny` = `reserve_curreny`

Reply:
It's updated to check the fee currency against the asset out currency, and consolidate the reserves when they match.

Reply:
Currently updating other reserve/unreserve locations to make sure correct currency type is reserved/unreserved when the outgoing currency matches the fees so the mechanism is always consistent (i.e. so we don't have any reserves done through native balances trait impl that are undone via orml impl.)

Reply:
All reserve/unreserve should be covered now, cleaning it up, adding additional coverage around this as well now.